snuba/migrations/runner.py

```diff
@@ -22,44 +25,70 @@ class Status(Enum):
         """
-        Run a single migration given its ID and marks the migration as complete.
+        Run a single migration given its migration key and marks the migration as complete.
         """
+        group = migration_key.group
+        migration_id = migration_key.migration_id
+
         assert all(
             cluster.is_single_node() for cluster in CLUSTERS
         ), "Cannot run migrations for multi node clusters"
 
-        logger.info(f"Running migration: {group} {migration_id}")
+        logger.info(f"Running migration: {migration_key}")
 
         migration = get_group_loader(group).load_migration(migration_id)
 
+        self._update_migration_status(migration_key, Status.IN_PROGRESS)
+
         operations = migration.forwards_local()
         for op in operations:
             op.execute()
 
-        logger.info(f"Finished running, updating status: {group} {migration_id}")
+        logger.info(f"Finished running, updating status: {migration_key}")
 
-        # TODO: In addition to marking migrations as completed, we should also mark
-        # migrations as in-progress before we execute the operations. However we
-        # will need to have some mechanism that allows this to be skipped in certain
-        # cases, such as the initial migration that creates the migrations table itself.
-        self._mark_completed(group, migration_id)
+        self._update_migration_status(migration_key, Status.COMPLETED)
 
-        logger.info(f"Finished: {group} {migration_id}")
+        logger.info(f"Finished: {migration_key}")
 
-    def _mark_completed(self, group: MigrationGroup, migration_id: str) -> None:
+    def _update_migration_status(
+        self, migration_key: MigrationKey, status: Status
+    ) -> None:
         statement = f"INSERT INTO {TABLE_NAME} FORMAT JSONEachRow"
         data = [
             {
-                "group": group.value,
-                "migration_id": migration_id,
+                "group": migration_key.group.value,
+                "migration_id": migration_key.migration_id,
                 "timestamp": datetime.now(),
-                "status": Status.COMPLETED.value,
+                "status": status.value,
             }
         ]
-        connection = get_cluster(StorageSetKey.MIGRATIONS).get_query_connection(
-            ClickhouseClientSettings.MIGRATE
-        )
-        connection.execute(statement, data)
+
+        try:
+            self.__connection.execute(statement, data)
+        except ClickhouseError as e:
+            # HACK: Ignore failures related to non existent table when we're
+            # setting system migrations to non completed statuses. This is required
+            # since the migrations table might not yet exist.
+            allow_fail = (
+                migration_key.group == MigrationGroup.SYSTEM
+                and status != Status.COMPLETED
+            )
+
+            if not allow_fail or e.code != errors.ErrorCodes.UNKNOWN_TABLE:
+                raise e
```

Comment:
I think there is a better way to do this. Options:
- move the running logic in the migrations parent class. Subclasses only define the steps. This way you can have a special migration class for the initial setup that does not have the same preconditions.
- Let a migration class provide its runner method/class, so that all migrations will provide the same runner class that does what this method does. The initial one will have its own runner that skips the initial state change.

I think it is worth not putting this special case in the runner (which is supposed to be generic) for a few reasons:
- it seems very ad hoc. If clickhouse changes error codes we are screwed.
- it makes the effort of representing system migrations as migration worthless, since then we need to treat them as special cases here.
- if all migrations work this way `pre conditions, set state, do stuff, set state` but there are some that do not, then the responsibility of deciding how each migration should work should be on the migration class, otherwise this is a leaky abstraction.

Both solutions should not be complex to implement.

Reply:
I moved the forwards method into the Migration class itself. I don't anticipate that any other migrations will need to override it apart from the first one though.